url_preview: Fix url-previews not shown for some webpages

Neeraj138 opened the pull request:
<!-- Describe your pull request here.-->
Fixes: #23664

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
- [x] Show title and description if found, even though an image isn't found.
![noimageurlpreview](https://user-images.githubusercontent.com/58552561/207035388-01df16f2-51ca-469f-9add-79da4569f5d3.png)
- [x] Support url preview for content-type text/xml. (Also, the image here is coming from twitter:image meta tag. So, the TwitterCard Parser that has been added as one of the fixed is also being tested in this example.)
![Screenshot 2022-12-19 00:30:00](https://user-images.githubusercontent.com/58552561/208314525-468a8779-d7d3-4b80-96ac-cdd3d90ff6ab.png)
![Screenshot 2022-12-18 23:57:38](https://user-images.githubusercontent.com/58552561/208314435-4142c3f5-6813-492a-8f7a-d8a198f7476a.png)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

Neeraj138 committed:
markdown-init: Show url preview when no image found.

Previously, the code checks if an image is found in the extracted data
for url preview. It doesn't show the url preview if no image is found
although there's a title and description. Fixed the behaviour to handle
this. Now, the url preview is shown even if the image is not found but
title/description have been found. Changed the test cases accordingly.

Fixes: a part of #23664

Neeraj138 committed:
url_preview: Add twitter meta tag parser.

Beside open graph(og) tags, many sites use twitter cards in meta tag.
We don't have a parser for this. So, added twitter_card.py which looks
for such tags. This parser has been implemented based on open_graph.py.

Fixes: part of #23664

Neeraj138 committed:
url_preview: Add support for text/xml content type.

Currently, url previews are only being supported for webpages
of content-type text/html only. This fix extends url preview
support to the webpages whose content-type is text/xml.

Fixes: a part of #23664

Co-authored by: @Beanpen

Neeraj138 commented:
> Since the description said development is in progress, I've converted this PR to a draft. Explicitly marking the PR as a draft helps reviewers know when a PR is not ready to be looked at yet.

Quoting the reply just to let you know that the PR is ready for review.

alya commented:
Nice! Works for me in light manual testing; I don't have an extensive list of sites to test.

alya commented:
@timabbott Could you please take a look and figure out next steps for this PR?

Neeraj138 committed:
url_preview: Fix no image after the first heading.

The old code only searches for the first image
after <h1> and uses it for the link preview.
This sometimes results in failures in
finding an image. This revision adds a search
to the siblings before h1 tag to look for an image.

Fixes: #23664

Co-authored by: @Beanpen
